proposals/0530-strict-box-denesting.rst

```diff
@@ -0,0 +1,486 @@
+and the aforementioned ``ShortByteString``) are essential parts of the
+user-facing API.
+
+
+GHC is already able to eliminate some packing and unpacking of these
+data types using less specific tools like case-of-known-constructor
+and the worker/wrapper transformation.  However, these tools share two
+limitations: A value can only be unpacked if it is used strictly and
+its type is specific enough to determine the relevant constructor.
+Notably, the latter implies that the contents of standard containers
+such as arrays, lists, sets, and tuples will never be unpacked.
+
+These limitations should be familiar: The same ones apply to unboxing
+``Int``\ s.  It is usually possible to work around these limitations,
+and libraries defining such types often strive to do so at their
+internal use-sites.  However, doing so can be a chore. The workarounds
+can have costs to code readability and are only effective when they
+are maintained.  (The workaround of monomorphising containers even has
+runtime cost, since it necessarily increases code size.)
+
+..
+  There are a few less essential limitations in today's GHC as well.
+  A constructor field of an unlifted type is always treated as a lazy
+  field and thus not unpacked, even with an explicit {-# UNPACK #-}
+  pragma.  Data types that capture an existential type variable are in
+  the same way ineligible for unpacking, and are also ignored by CPR
+  analysis.
+
+..
+  TODO: perhaps add links to past conversations demonstrating demand
+
+
+Proposed Change Specification
+-----------------------------
+..
+  This section shares a lot with the introduction.
+  When editing one, it is wise to consider editing the other!
+
+This proposal introduces a new optimization flag, tentatively named
+``-fstrict-box-denesting``, which will be enabled by default with
+``-O1``.  When this flag is active while compiling a ``data`` or
+``data instance`` declaration, its constructors are tested for
+"denestability." A constructor is considered denestable when it
+satisfies these four conditions:
+
+..
+  These are identical to the four conditions given in the introduction.
+
+1. There is no other constructor in the relevant ``data`` or
+   ``data instance`` declaration.
+2. Among the constructor's fields (after unpacking, counting
+   existentials and constraints), exactly one field has
+   non-zero-width.
+3. That field's type is represented by a GC-managed pointer.
+4. If that field is of lifted type, it must be a strict field.
+
+A constructor determined to be denestable will not generate any code
+at any of its use sites (even use-sites in modules for which
+``-fstrict-box-denesting`` is disabled), so that using or
+pattern-matching against this constructor has no runtime cost, except
+to the extent that doing so forces evaluation of the field.
+
+Whether a constructor is determined denestable or not has no effect on
+the surface-language semantics of that constructor.  In particular,
```

Comment:
If it doesn't change surface language, I think it may not strictly need to go through the proposal process. But of course we can go through it, maybe some useful feedback is gathered this way.

Very nicely written, btw.

Reply:
Thank you. Drafting a proposal seemed much more clearly appropriate before I realized that giving `dataToTag#` a constraint-evidence argument can sidestep the risk of silently subtly breaking some machine-generated code using it somewhere.

But this also seemed like a better place to lay out my envisioned roadmap an seek discussion than reusing [the old ghc wiki page on this potential optimization](https://gitlab.haskell.org/ghc/ghc/-/wikis/newtype-optimization-for-gadts), whose name I don't like very much and whose working conventions I remain ignorant of.